crates/sui-types/src/object.rs

```diff
@@ -398,6 +400,68 @@ impl MoveObject {
     }
 }
 
+// Helpers for extracting Coin<T> balances for all T
+impl MoveObject {
+    fn is_balance(s: &StructTag) -> Option<&TypeTag> {
+        (Balance::is_balance(s) && s.type_params.len() == 1).then_some(&s.type_params[0])
+    }
+
+    /// Get the total balances for all Coin<T> embedded in `self`.
+    pub fn get_coin_balances(
+        &self,
+        layout_resolver: &mut dyn LayoutResolver,
+    ) -> Result<HashMap<TypeTag, u64>, SuiError> {
+        let mut balances = HashMap::default();
+
+        // Fast path without deserialization.
+        if let Some(type_tag) = self.type_.coin_type_maybe() {
+            let balance = self.get_coin_value_unsafe();
+            *balances.entry(type_tag).or_insert(0) += balance;
+        } else {
+            let layout = layout_resolver.get_layout(self, ObjectFormatOptions::with_types())?;
+            let move_struct = self.to_move_struct(&layout)?;
+            Self::get_coin_balances_in_struct(&move_struct, &mut balances);
+        }
+
+        Ok(balances)
+    }
+
+    /// Get the total balances for all Coin<T> embedded in `s`, eitehr directly or in its
+    /// (transitive fields).
+    fn get_coin_balances_in_struct(s: &MoveStruct, balances: &mut HashMap<TypeTag, u64>) {
+        let (struct_type, fields) = match s {
+            MoveStruct::WithTypes { type_, fields } => (type_, fields),
+            _ => unreachable!(),
+        };
+
+        if let Some(type_tag) = Self::is_balance(struct_type) {
+            let balance = match fields[0].1 {
+                MoveValue::U64(n) => n,
+                _ => unreachable!(), // a Balance<T> object should have exactly one field, of type int
+            };
+
+            // Accumulate the found balance
+            *balances.entry(type_tag.clone()).or_insert(0) += balance;
+        } else {
+            for field in fields {
+                Self::get_coin_balances_in_value(&field.1, balances);
```

Comment:
As @tzakian can attest, we've had some "fun" security fixes related to using functions that recurse through a  maliciously crafted Move value without tracking the depth of the recursion. We probably want to use the same sort of depth-limiting pattern here to protect ourselves (and in `get_total_sui`, if we don't already).

Reply:
Ok i added some recursion limiting to 2x what the protocol config allows currently (so 256 here vs 128 in the protocol config). Please double check my logic to make sure i'm counting properly, unless the intent is just to ensure that we have *some* limit.

I also just rewrote `get_total_sui` to internally use this new `get_total_balances` function for maintainability